From b41c9b0266e8370033a7799f6806bfc70b7fd75f Mon Sep 17 00:00:00 2001
From: Eric Wheeler <bcache@linux.ewheeler.net>
Date: Fri, 13 Oct 2017 16:35:33 -0700
Subject: [PATCH] bcache: update bio->bi_opf bypass/writeback REQ_ flag hints
Git-commit: b41c9b0266e8370033a7799f6806bfc70b7fd75f
Patch-mainline: v4.15-rc1
References: bsc#1076110

Flag for bypass if the IO is for read-ahead or background, unless the
read-ahead request is for metadata (eg, from gfs2).
        Bypass if:
                bio->bi_opf & (REQ_RAHEAD|REQ_BACKGROUND) &&
			!(bio->bi_opf & REQ_META))

        Writeback if:
                op_is_sync(bio->bi_opf) ||
			bio->bi_opf & (REQ_META|REQ_PRIO)

(Coly Li: rebased for SLE12-SP2)

Signed-off-by: Eric Wheeler <bcache@linux.ewheeler.net>
Reviewed-by: Coly Li <colyli@suse.de>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Coly Li <colyli@suse.de>

---
 drivers/md/bcache/request.c   |    8 ++++++++
 drivers/md/bcache/writeback.h |    5 +++--
 2 files changed, 11 insertions(+), 2 deletions(-)

--- a/drivers/md/bcache/request.c
+++ b/drivers/md/bcache/request.c
@@ -384,6 +384,14 @@ static bool check_should_bypass(struct c
 	     (bio->bi_rw & REQ_WRITE)))
 		goto skip;
 
+	/*
+	 * Flag for bypass if the IO is for read-ahead or background,
+	 * unless the read-ahead request is for metadata (eg, for gfs2).
+	 */
+	if (bio->bi_rw & (REQ_RAHEAD) &&
+	    !(bio->bi_rw & REQ_META))
+		goto skip;
+
 	if (bio->bi_iter.bi_sector & (c->sb.block_size - 1) ||
 	    bio_sectors(bio) & (c->sb.block_size - 1)) {
 		pr_debug("skipping unaligned io");
--- a/drivers/md/bcache/writeback.h
+++ b/drivers/md/bcache/writeback.h
@@ -76,8 +76,9 @@ static inline bool should_writeback(stru
 	if (would_skip)
 		return false;
 
-	return bio->bi_rw & REQ_SYNC ||
-		in_use <= CUTOFF_WRITEBACK;
+	return (bio->bi_rw & REQ_SYNC ||
+		bio->bi_rw & (REQ_META|REQ_PRIO) ||
+		in_use <= CUTOFF_WRITEBACK);
 }
 
 static inline void bch_writeback_queue(struct cached_dev *dc)
